rust/worker/src/execution/orchestration/compact.rs

```diff
@@ -286,31 +333,44 @@ impl CompactOrchestrator {
             }
         };
 
-        self.num_write_tasks = partitions.len() as i32;
-        for parition in partitions.iter() {
-            let operator = WriteSegmentsOperator::new();
-            let input = WriteSegmentsInput::new(
+        self.num_write_tasks = partitions.len() as i32 * 3;
+        for partition in partitions.iter() {
+            // We must have an atomic that tracks the current offset ID across partitions to guarantee unique offset IDs for every record.
+            // Every partition has a log materializer that is responsible for updating this atomic.
+            // However, because the log materializer returns borrowed values, we currently run the materializer multiple times for each partition (for each segment writer).
+            // This means that only one log materializer run per partition should access and update the atomic shared across all partitions, and the atomics for the other log materializer runs should not be shared.
```

Comment:
This is going to be the hardest thing to debug and prove.  I'm just spitballing:  Can we make a consumable type that enforces this property?

Reply:
discussed in-person, the correct solution is to run the log materializer once (not easy without further refactoring)

Reply:
Is there any other solution here - this is super hairy / iffy

Reply:
I thought about it for a while and I didn't see any other options. Agree it's hacky, open to something better.

Reply:
I went through this today to double-check the logic and it's actually incorrect. Consider this case:
- there are two partitions, materializing logs in two separate threads
- the first partition handles the first 3 records (indices 0-2, inclusive)
- the second partition handles the next 2 records (indices 3-4, inclusive)
- the secondary materialization, using the unshared atomic, runs either in one of the two threads or somewhere else--it doesn't matter

We can then have this case:

```
Thread 1 (partition 1, #0-2)             thread 2 (partition 2, #3-4        offset_shared unshared_offset_thread_1
                                                                             0             0
Primary mat increments offset (#0)                                           1             0
                                         Primary mat increments offset (#3)  2             0
                                         Primary mat increments offset (#4)  3             0
Primary mat increments offset (#1)                                           4             0
Secondary mat increments offset (#0)                                         3             1
Secondary mat increments offset (#1)                                         3             2
```

The expected mapping (record index -> offset ID) is:

- 0 -> 1
- 1 -> 2
- 2 -> 4
- 3 -> 3

However, the secondary materializer with the unshared atomic ends up mapping 1 -> 2.
-----------------------

I will investigate and see how much work it would be to refactor materialization so that it can be run in an operator once.

Reply:
I have a refactor here that puts materialization inside an operator: https://github.com/chroma-core/chroma/pull/3166. I'll restack this PR on top of that stack.